File: e-dashboard-api/src/Domain/Diagram/Shapes/ConnectShape.cs
using System.Text;

namespace EnergyDashboard.Domain.Diagram.Shapes;

public interface IConnectShapeProperties : IProperties
{
    string SrcShapeName { get; set; }
    string DesShapeName { get; set; }
}

public class ConnectShapeProperties : IConnectShapeProperties
{
    public string Name { get; set; }
    public double Voltage { get; set; }
    public string SrcShapeName { get; set; }
    public string DesShapeName { get; set; }
    public ConnectShapeProperties()
    {
        this.SrcShapeName = string.Empty;
        this.DesShapeName = string.Empty;
    }
}

public class ConnectShape : LineConnectShape
{
    public ConnectShape() : base()
    {
        this.Properties = new ConnectShapeProperties();
        this.Shape = new Line(new Point(0, 0), new Point(0, 0));
    }
    public ConnectShape(string name) : base(name)
    {
        this.Properties = new ConnectShapeProperties();
        this.Shape = new Line(new Point(0, 0), new Point(0, 0));
    }

    public new ConnectShapeProperties Properties { get; set; }

    public override string toSVG()
    {
        StringBuilder markup = new StringBuilder();
        var s = this.Shape as Line;

        markup.Append(
            "\t\t<line id=\"" + this.Id + "\"" +
            " x1=\"" + s.P1.X + "\"" +
            " y1=\"" + s.P1.Y + "\"" +
            " x2=\"" + s.P2.X + "\"" +
            " y2=\"" + s.P2.Y + "\"" +
            " style=\"" + Util.Utilility.GetSvgStyles(this.Style) + "\"/>\n");

        return markup.ToString();
    }
}

File: e-dashboard-ui/src/ClientApp/src/views/Diagram/DiagramLib/Shapes/ConnectShape.ts
import { IPoint } from './interfaces/IShapes';
import { LineConnectShape } from './LineConnectShape';
import { DrawingToolType } from '../Enums';
import { SVGUtil } from '../Util/SVGUtil';
import { IProperties } from './interfaces';
import { ISchema } from './interfaces/ISchema';
import { Line, Point } from './Shapes';

interface IConnectShapeProperties extends IProperties {
    srcShapeName: string | undefined;
    desShapeName: string | undefined;
}

export class ConnectShapeProperties implements IConnectShapeProperties {
    public name: string;
    public voltage:number;
    public srcShapeName: string | undefined;
    public desShapeName: string | undefined;


    constructor(name: string) {
        this.name = name;
        this.voltage = 0;
        this.srcShapeName = undefined;
        this.desShapeName = undefined;
    }
}

export class ConnectShape extends LineConnectShape<Line> {
    protected properties : IConnectShapeProperties;
    
    constructor(name: string, x: number, y: number) {
        super(name);
        this.$type = "ConnectShape";
        this.shape = new Line(new Point(x, y), new Point(x+1, y+1));
        this.properties = new ConnectShapeProperties(name);
    }

    public get location(): IPoint {
        return new Point(this.shape.p1.x, this.shape.p1.y);
    }

    public draw(ctx: CanvasRenderingContext2D) {
        ctx.lineWidth = this.style.strokeWidth;
        ctx.globalAlpha = this.style.opacity;
        ctx.strokeStyle = this.style.strokeStyle;

        ctx.beginPath();
        ctx.moveTo(this.shape.p1.x, this.shape.p1.y);
        ctx.lineTo(this.shape.p2.x, this.shape.p2.y);
        ctx.stroke();
        this._drawHandle(ctx);
    }
    public move(to: IPoint) {
    }
    public inResizeZone(point: IPoint): number {
        return -1;
    }
    public resizeToLocation(to: IPoint, handle: number) {
        if (handle === 0) {
            this.shape.p1.x = to.x;
            this.shape.p1.y = to.y;
        }
        else {
            this.shape.p2.x = to.x;
            this.shape.p2.y = to.y;
        }
    }

    public contains(mousePoint: IPoint): boolean {
        return this.shape.contains(mousePoint);
    }
    public getMoveOffset(mousePosition: IPoint): IPoint {
        return new Point(mousePosition.x - this.shape.p1.x, mousePosition.y - this.shape.p1.y);
    }
    public getCursorType(mousePoint: IPoint, tool: DrawingToolType): string {
        if (this.inResizeZone(mousePoint) === 0)
            return "w-resize";
        else if (this.inResizeZone(mousePoint) === 1)
            return "e-resize";
        else
            return "move";
    }

    public toSVG(): string {
        var markup: Array<string> = [];
        markup.push(
            '\t<line ', this.getSvgId(),
            'x1="', this.shape.p1.x + '"',
            ' y1="', this.shape.p1.y + '"',
            ' x2="', this.shape.p2.x + '"',
            ' y2="', this.shape.p2.y + '"',
            ' style="', SVGUtil.getSvgStyles(this.style),
            '"/>\n'
        );
        return markup.join('');
    }

    get Schema(): ISchema {
        let schema: ISchema = {
            title: "ویژگی پیوند",
            type: "object",
            tabs: [
                {
                    tabIndex: 1,
                    tabName: 'عمومی',
                    properties: {
                        "name": {
                            "type": "string",
                            "title": "نام",
                            "disabled": true,
                        },
                        "srcShapeName": {
                            "type": "string",
                            "title": "مبدا",
                            "disabled": true,
                        },
                        "desShapeName": {
                            "type": "string",
                            "title": "مقصد",
                            "disabled": true,
                        },
                    }
                }
            ]
        };

        return schema;
    }

    public get Properties(): IConnectShapeProperties {
        return this.properties;
    }

    public set Properties(prop: IConnectShapeProperties) {
        this.properties = prop;
        this.voltage = prop.voltage;
        this.name = prop.name;
    }

    protected _drawHandle(ctx: CanvasRenderingContext2D) {
        if (this.isSelected) {
            var d = this.selectionZoneWidth;
            ctx.globalAlpha = this.style.opacity;
            ctx.strokeStyle = "#00FFFF";
            ctx.lineWidth = 1;
            ctx.strokeRect(this.shape.p1.x - d, this.shape.p1.y - 2, 2 * d, 2 * d);
            ctx.strokeRect(this.shape.p2.x - d, this.shape.p2.y - 2, 2 * d, 2 * d);
        }
    }
}
